// @jnxplus/gradle/src/generators/application/generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const path = require("path");
const common_1 = require("@jnxplus/common");
const _1 = require("../../.");
const fs = require("fs");
function normalizeOptions(plugin, tree, options) {
    const simpleProjectName = (0, devkit_1.names)((0, common_1.normalizeName)(options.name)).fileName;
    let projectName;
    if (options.simpleName) {
        projectName = simpleProjectName;
    }
    else {
        projectName = options.directory
            ? `${(0, common_1.normalizeName)((0, devkit_1.names)(options.directory).fileName)}-${simpleProjectName}`
            : simpleProjectName;
    }
    const projectDirectory = options.directory
        ? `${(0, devkit_1.names)(options.directory).fileName}/${simpleProjectName}`
        : simpleProjectName;
    const projectRoot = `${(0, devkit_1.getWorkspaceLayout)(tree).appsDir}/${projectDirectory}`;
    const parsedTags = options.tags
        ? options.tags.split(',').map((s) => s.trim())
        : [];
    let appClassName = '';
    if (plugin === '@jnxplus/nx-micronaut-gradle' ||
        options.framework === 'micronaut') {
        appClassName = (0, devkit_1.names)(projectName).className;
    }
    else {
        appClassName = `${(0, devkit_1.names)(projectName).className}Application`;
    }
    let packageName;
    if (options.simplePackageName) {
        packageName = `${options.groupId}.${(0, devkit_1.names)(simpleProjectName).className.toLocaleLowerCase()}`.replace(new RegExp(/-/, 'g'), '');
    }
    else {
        packageName = `${options.groupId}.${options.directory
            ? `${(0, devkit_1.names)(options.directory).fileName.replace(new RegExp(/\//, 'g'), '.')}.${(0, devkit_1.names)(simpleProjectName).className.toLocaleLowerCase()}`
            : (0, devkit_1.names)(simpleProjectName).className.toLocaleLowerCase()}`.replace(new RegExp(/-/, 'g'), '');
    }
    const packageDirectory = packageName.replace(new RegExp(/\./, 'g'), '/');
    const linter = options.language === 'java' ? 'checkstyle' : 'ktlint';
    const isCustomPort = !!options.port && +options.port !== 8080;
    const dsl = (0, _1.getDsl)(tree);
    const kotlinExtension = dsl === 'kotlin' ? '.kts' : '';
    let qVersion = '';
    if (plugin === '@jnxplus/nx-quarkus-gradle' ||
        options.framework === 'quarkus') {
        const gradlePropertiesPath = path.join(devkit_1.workspaceRoot, 'gradle.properties');
        const gradlePropertiesContent = fs.readFileSync(gradlePropertiesPath, 'utf-8');
        qVersion = (0, _1.getQuarkusVersion)(gradlePropertiesContent);
        if (common_1.quarkusVersion === undefined) {
            qVersion = common_1.quarkusVersion;
        }
    }
    return Object.assign(Object.assign({}, options), { projectName,
        projectRoot,
        projectDirectory,
        parsedTags,
        appClassName,
        packageName,
        packageDirectory,
        linter,
        isCustomPort,
        dsl,
        kotlinExtension, quarkusVersion: qVersion });
}
function addFiles(d, plugin, tree, options) {
    if (plugin === '@jnxplus/nx-boot-gradle' ||
        options.framework === 'spring-boot') {
        addBootFiles(d, tree, options);
    }
    if (plugin === '@jnxplus/nx-quarkus-gradle' ||
        options.framework === 'quarkus') {
        addQuarkusFiles(d, tree, options);
    }
    if (plugin === '@jnxplus/nx-micronaut-gradle' ||
        options.framework === 'micronaut') {
        addMicronautFiles(d, tree, options);
    }
    if (options.framework === 'none') {
        addNoneFiles(d, tree, options);
    }
}
function addNoneFiles(d, tree, options) {
    const templateOptions = Object.assign(Object.assign(Object.assign({}, options), (0, devkit_1.names)(options.name)), { offsetFromRoot: (0, devkit_1.offsetFromRoot)(options.projectRoot), template: '' });
    (0, devkit_1.generateFiles)(tree, path.join(d, 'files', 'none', options.language), options.projectRoot, templateOptions);
    const fileExtension = options.language === 'java' ? 'java' : 'kt';
    if (options.minimal) {
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/main/${options.language}/${options.packageDirectory}/App.${fileExtension}`));
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/test/${options.language}/${options.packageDirectory}/AppTest.${fileExtension}`));
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/test/resources/application${options.configFormat}`));
    }
}
function addBootFiles(d, tree, options) {
    const templateOptions = Object.assign(Object.assign(Object.assign({}, options), (0, devkit_1.names)(options.name)), { offsetFromRoot: (0, devkit_1.offsetFromRoot)(options.projectRoot), template: '' });
    (0, devkit_1.generateFiles)(tree, path.join(d, 'files', 'boot', options.language), options.projectRoot, templateOptions);
    const fileExtension = options.language === 'java' ? 'java' : 'kt';
    if (options.packaging === 'jar') {
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/main/${options.language}/${options.packageDirectory}/ServletInitializer.${fileExtension}`));
    }
    if (options.minimal) {
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/main/${options.language}/${options.packageDirectory}/HelloController.${fileExtension}`));
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/test/${options.language}/${options.packageDirectory}/HelloControllerTests.${fileExtension}`));
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/test/resources/application${options.configFormat}`));
        if (options.language === 'kotlin') {
            tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, '/src/test/resources/junit-platform.properties'));
        }
    }
}
function addQuarkusFiles(d, tree, options) {
    const templateOptions = Object.assign(Object.assign(Object.assign({}, options), (0, devkit_1.names)(options.name)), { offsetFromRoot: (0, devkit_1.offsetFromRoot)(options.projectRoot), template: '' });
    (0, devkit_1.generateFiles)(tree, path.join(d, 'files', 'quarkus', options.language), options.projectRoot, templateOptions);
    if (options.minimal) {
        const fileExtension = options.language === 'java' ? 'java' : 'kt';
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/main/${options.language}/${options.packageDirectory}/GreetingResource.${fileExtension}`));
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/test/${options.language}/${options.packageDirectory}/GreetingResourceTest.${fileExtension}`));
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/native-test/${options.language}/${options.packageDirectory}/GreetingResourceIT.${fileExtension}`));
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/main/resources/META-INF/resources/index.html`));
    }
    else {
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/main/${options.language}/.gitkeep`));
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/test/${options.language}/.gitkeep`));
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/native-test/${options.language}/.gitkeep`));
    }
}
function addMicronautFiles(d, tree, options) {
    const templateOptions = Object.assign(Object.assign(Object.assign({}, options), (0, devkit_1.names)(options.name)), { offsetFromRoot: (0, devkit_1.offsetFromRoot)(options.projectRoot), template: '' });
    (0, devkit_1.generateFiles)(tree, path.join(d, 'files', 'micronaut', options.language), options.projectRoot, templateOptions);
    if (options.minimal) {
        const fileExtension = options.language === 'java' ? 'java' : 'kt';
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/main/${options.language}/${options.packageDirectory}/HelloController.${fileExtension}`));
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/test/${options.language}/${options.packageDirectory}/HelloControllerTest.${fileExtension}`));
    }
    else {
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/main/${options.language}/.gitkeep`));
        tree.delete((0, devkit_1.joinPathFragments)(options.projectRoot, `/src/test/${options.language}/.gitkeep`));
    }
}
function default_1(d, plugin, tree, options) {
    var _a;
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const normalizedOptions = normalizeOptions(plugin, tree, options);
        const projectConfiguration = {
            root: normalizedOptions.projectRoot,
            projectType: 'application',
            sourceRoot: `${normalizedOptions.projectRoot}/src`,
            targets: {
                build: {
                    executor: `${plugin}:build`,
                    outputs: [`${normalizedOptions.projectRoot}/build`],
                },
                'build-image': {},
                serve: {},
                lint: {
                    executor: `${plugin}:lint`,
                    options: {
                        linter: `${normalizedOptions.linter}`,
                    },
                },
                test: {
                    executor: `${plugin}:test`,
                },
                ktformat: {},
            },
            tags: normalizedOptions.parsedTags,
        };
        const targets = (_a = projectConfiguration.targets) !== null && _a !== void 0 ? _a : {};
        if (plugin === '@jnxplus/nx-boot-gradle' ||
            options.framework === 'spring-boot') {
            targets['build'].options = Object.assign(Object.assign({}, targets['build'].options), { packaging: `${normalizedOptions.packaging}` });
        }
        if (options.framework === 'none') {
            targets['serve'] = {
                executor: `${plugin}:run-task`,
                options: {
                    task: 'run',
                },
            };
        }
        if (options.framework !== 'none') {
            targets['build-image'] = {
                executor: `${plugin}:build-image`,
            };
            targets['serve'] = {
                executor: `${plugin}:serve`,
            };
        }
        if (options.framework && options.framework !== 'none') {
            targets['build'].options = Object.assign(Object.assign({}, targets['build'].options), { framework: options.framework });
            targets['build-image'].options = Object.assign(Object.assign({}, targets['build-image'].options), { framework: options.framework });
            targets['serve'].options = Object.assign(Object.assign({}, targets['serve'].options), { framework: options.framework });
        }
        if (options.language === 'kotlin') {
            targets['ktformat'] = {
                executor: `${plugin}:ktformat`,
            };
        }
        (0, common_1.clearEmpties)(targets);
        (0, devkit_1.addProjectConfiguration)(tree, normalizedOptions.projectName, projectConfiguration);
        addFiles(d, plugin, tree, normalizedOptions);
        (0, _1.addProjectToGradleSetting)(tree, normalizedOptions);
        yield (0, devkit_1.formatFiles)(tree);
    });
}
exports.default = default_1;
